### ISSUE TITLE
wire: provide an example of mocking

### ISSUE THREAD

**Issue by gracenoah** [google/go-cloud#477]
It's easy to see how one would use wire in the main.go of an executable to wire up everything it needs, but it's not clear to me how one would go about writing tests with wire. This is simlar to https://github.com/google/go-cloud/issues/223, but in the case of tests, it might be necessary to dynamically construct an object with a mock inserted in an arbitrary point in the graph, potentially dynamically. What's a the recommended way of doing this? Would you just expose the entire graph to the test after it's built so it can take what it needs and manually reassemble new objects from it?

Let me give a concrete example:

```go
type Hello struct {}

func (h Hello) Greet() string {
    return "Hello world"
}

type Greeter interface {
    Greet() string
}

func ProvideGreeter() Greeter {
    return Hello{}
}

type App struct {
    Greeter Greeter
    Timer Timer
}

type RealTimer struct {}

func (r RealTimer) Now() string {
    return time.Now().String()
}

type Timer interface {
    Now() string
}

func ProvideTimer() Timer {
    return RealTimer{}
}

func NewApp(g Greeter, t Timer) App {
    return App{Greeter: g, Timer: t}
}

func (a App) GreetNow() string {
    return a.Timer.Now() + " " + a.Greeter.Greet()
}
```

Now let's say you have an injector like this that you normally use to construct your app:
```go
func InitializeApp(ctx context.Context) (App, error) {
	wire.Build(ProviderGreeter, ProvideTimer)
	return App{}, nil
}
```

Now you want to mock out time in a test, but use an otherwise normal app:
```go
type MockTimer struct {
    Time string
}

func (m MockTimer) Now() string {
    return m.Time
}
```

Ideas for how you might do that:

- You can create a different initializer for your test, but if you do that, you can't dynamically configure the mock based on ex. a table driven test. If you end up using globals to configure the mock, you might also run into issues with concurrent tests.
- You can use the real initializer, create a full App, then reach in and swap out the implementation of the Timer before calling any methods on the App
- You can use the real initializer to create a full App, then reach in and use the Greeter and your own Timer to create a new App. This is easy only because App is the top level of the tree, but if it's deeply nested, you'd have to rebuild everything above it manually.

Is there a better way to do mocking? Am I missing another good option?

**Comment by vangent** [comment]
I think you'll want your first option, a different initializer for your test. You could pass it the things you want to mock:

```
func InitializeAppWithMocks(ctx context.Context, mockTimer Timer) (App, error) {
	wire.Build(ProvideGreeter)
	return App{}, nil
}
```

or even the args needed to make the mocks, although this will get painful if there are many:

```
func InitializeAppWithMocks(ctx context.Context, mockTime string) (App, error) {
	wire.Build(ProvideGreeter, MockTimer{})
	return App{}, nil
}
```

Or, you could have it return more of what it creates so you can modify them later:

```
func ProvideEmptyMockTimer() MockTimer { return MockTimer{} }

func InitializeAppWithMocks(ctx context.Context) (App, MockTime, error) {
	wire.Build(ProvideGreeter, ProvideEmptyMockTimer)
	return App{}, MockTime{}, nil
}
```

**Comment by gracenoah** [comment]
I didn't know you can have arbitrary arguments to initializers that replace providers. Looking at the docs again, it explains the context passing, but I didn't understand it the first time I read it. The first option seems the simplest right now. The only down side is that I'd have to make a provider set that excludes the Timer. That's where "subtract" https://github.com/google/go-cloud/issues/51 would come in.

I don't understand second example. Can you clarify that one? MockTimer{} is used as a provider and a string is passed in as an argument? How would wire know to create the mock timer with that string?

The third example doesn't seem to work. Wire complains that the second return value is not a function. Maybe it would work if the return value of the initializer was just one struct that contains both the app and the mock. In that case, I think this is similar to idea 2, but instead of the app exposing its internals and reaching into the app to replace the timer, you expose the timer yourself in the test and reach into it to configure it. Let's call this idea 4. 

I like this the most so far and it would work well with passing in a *testing.T into the initializer, but again, I wish there was a better way to create the right provider set. In this case it would be a `replace` instead of a `subtract`.

**Comment by vangent** [comment]
In the second example, MockTimer{} is being used as a [struct provider](https://github.com/google/go-cloud/tree/master/wire#struct-providers). Wire constructs it by finding values for all of the struct fields, use the type to find a provider. This is similar to how `App` is being created in your example; there's no explicit `ProvideApp` provider, it's just finding values for the types of the struct fields. So, to construct the MockTimer{} struct, Wire will look for a string to match the `Time string` field, and it will find the mockTime string.

If you have multiple mocks with multiple args, the types will likely overlap (e.g., mock1 and mock2 might both need strings). You could get it to work by defining more specific types (e.g., `type MyMockTime string`), but it might be painful.

You are right on the third example, you'd need to wrap App and MockTime in a struct. Sorry about that.

Note that #51 is suggesting a command line tool that outputs code, not a runtime or even `generate` time thing.

You may end up with two ProviderSets, something like this:

```
var RealSet = wire.NewSet(ProvideGreeter, ProvideTimer)
var MockSet = wire.NewSet(ProvideGreeter, ProvideEmptyMockTimer)

func InitializeApp(ctx context.Context) (App, error) {
	wire.Build(RealSet)
	return App{}, nil
}

type AppMocks struct {
  App App
  MockTimer MockTimer
}

func InitializeAppWithMocks(ctx context.Context) (AppWithMocks,  error) {
	wire.Build(MockSet)
	return App{}, nil
}
```

**Comment by vangent** [comment]
On reviewing this, I think that actually my last suggestion won't work. `ProviderEmptyMockTimer` returns either a `MockTimer` or a `Timer`, and while `wire` supports using a value as an interface that it implements, I don't think it supports using it as both the concrete type and the interface.

I'm adding an example to the `wire` tests, and will check with @zombiezen to see if I'm missing something.

**Comment by vangent** [comment]
@zombiezen pointed out that `wire.Bind` resolves the issue I pointed out earlier.

So, you can in fact do my last suggestion, you just need to use `wire.Bind` to bind the concrete mock type to the interface. Please see the in-flight PR for documentation and an example of both approaches.

**Comment by vangent** [comment]
Fixed in PR #488.
